# Naive Bayes Bigram Pipeline

In [2]:
print(true_df.columns)

Index(['title', 'text', 'subject', 'date'], dtype='object')


In [3]:
# Import the Naive Bayes model and dataset
from models.naive_bayes_bigram_model import NaiveBayesBigram
from utils.utils import fake_df

# Initialize the Naive Bayes model
nb_model = NaiveBayesBigram()

# Train the model on the dataset
# Assuming `text` is the column with text data and `label` is the column with labels
results = nb_model.train(fake_df, text_column='text', label_column='subject')

# Print the results
print(f"Naive Bayes Accuracy: {results['accuracy']:.4f}")
print("\nClassification Report:")
print(results['classification_report'])

Naive Bayes Accuracy: 0.4603

Classification Report:
                 precision    recall  f1-score   support

Government News       0.08      0.10      0.09       316
    Middle-east       0.12      0.14      0.13       159
           News       0.84      0.91      0.87      1821
        US_News       0.08      0.08      0.08       160
      left-news       0.13      0.12      0.12       897
       politics       0.28      0.24      0.26      1344

       accuracy                           0.46      4697
      macro avg       0.26      0.27      0.26      4697
   weighted avg       0.44      0.46      0.45      4697



In [1]:
# Import necessary libraries
import pandas as pd
from models.naive_bayes_bigram_model import NaiveBayesBigram

# Load the dataset from a CSV file
file_path = "/Users/haigbedros/Desktop/MSDS/Capstone/CODE/ml-models-information-filtering/notebooks/lda_bert_processed_marr.csv"  # Replace with the path to your CSV file
df = pd.read_csv(file_path)

# Initialize the Naive Bayes model
nb_model = NaiveBayesBigram()

# Train the model on the dataset
# Assuming `text` is the column with text data and `label` is the column with labels
results = nb_model.train(df, text_column='text', label_column='label')

# Print the results
print(f"Naive Bayes Accuracy: {results['accuracy']:.4f}")
print("\nClassification Report:")
print(results['classification_report'])

KeyError: 'label'

Analysis of Results
Class Imbalance:

The Classification Report shows that the News class has significantly higher support (1821 samples) compared to other classes like Middle-east (159 samples) or US_News (160 samples).
This imbalance likely skews the model's predictions toward the majority class (News), as evidenced by its high precision, recall, and F1-score.
Poor Performance on Minority Classes:

Classes like Government News, Middle-east, and US_News have very low precision, recall, and F1-scores, indicating that the model struggles to correctly classify these categories.
Weighted Average:

The weighted average F1-score is 0.45, which reflects the model's overall performance across all classes, weighted by the number of samples in each class.
Naive Bayes Limitations:

Naive Bayes assumes that features (bigrams in this case) are conditionally independent, which may not hold true for text data.
The model may not capture the contextual relationships between words as effectively as more advanced models like BERT.

Future: we will use techniques like SMOTE (Synthetic Minority Oversampling Technique) or random oversampling to balance the dataset.


**Then will compare with Advanced Models**

LDA + BERT Hybrid:
Compare the Naive Bayes results with your LDA + BERT pipeline to validate your hypothesis that the hybrid approach will outperform simpler models like Naive Bayes.

**Lastly will use Dimensionality Reduction for Visualization**

Visualize the bigram features using dimensionality reduction techniques like PCA or t-SNE to understand how well the classes are separated in the feature space.
